nixos/modules/services/misc/snapper.nix

```diff
@@ -3,7 +3,101 @@
 with lib;
 
 let
+  inherit (lib.strings) concatMapStrings concatMapStringsSep escape removeSuffix toUpper;
+
   cfg = config.services.snapper;
+
+  camelToCaps = with builtins; s: toUpper (concatMapStrings (p: if isList p then "${elemAt p 0}_${elemAt p 1}" else p) (split "([^A-Z])([A-Z])" s));
+
+  mkValue = with builtins; v:
+    if isList v then "\"${concatMapStringsSep " " (escape [ "\\" " " ]) v}\""
+    else if v == true then "yes"
+    else if v == false then "no"
+    else if isString v then "\"${v}\""
+    else toJSON v;
+
+  mkKeyValue = k: v: "${camelToCaps k}=${mkValue v}";
+
+  # "it's recommended to always specify the filesystem type"  -- man snapper-configs
+  defaultOf = k: if k != "fstype" && configOptions ? ${k}.default then configOptions.${k}.default else null;
+
+  safeStr = types.strMatching "[^\n\"]*" // {
+    description = "string without line breaks or quotes";
+    descriptionClass = "conjunction";
+  };
+
+  configOptions = {
+    subvolume = mkOption {
+      type = types.path;
+      description = lib.mdDoc ''
+        Path of the subvolume or mount point.
+        This path is a subvolume and has to contain a subvolume named
+        .snapshots.
+        See also man:snapper(8) section PERMISSIONS.
+      '';
+    };
+
+    fstype = mkOption {
+      type = types.enum [ "btrfs" ];
+      default = "btrfs";
+      description = lib.mdDoc ''
+        Filesystem type. Only btrfs is stable and tested.
+      '';
+    };
+
+    allowGroups = mkOption {
+      type = types.listOf safeStr;
+      default = [];
+      description = lib.mdDoc ''
+        List of groups allowed to operate with the config.
+
+        Also see the PERMISSIONS section in man:snapper(8).
+      '';
+    };
+
+    allowUsers = mkOption {
+      type = types.listOf safeStr;
+      default = [];
+      example = [ "alice" ];
+      description = lib.mdDoc ''
+        List of users allowed to operate with the config. "root" is always
+        implicitly included.
+
+        Also see the PERMISSIONS section in man:snapper(8).
+      '';
+    };
+
+    timelineCleanup = mkOption {
+      type = types.bool;
+      default = false;
+      example = true;
```

Comment:
In my opinion this doesn't really add any value. Maybe other people have different opinions, though. :man_shrugging:

Reply:
Heh, I was just inlining what `mkEnableOption` does aside from the description, which I assumed was written to do the best thing. Happy to follow whatever the best practice actually is for boolean options.